### PR TITLE
Upgrade `resolvelib>=0.8.1,<0.9.0`

### PR DESCRIPTION
**Pull request by hswong3i**
##### SUMMARY

Since 2021-09-27 `resolvelib` author comment with "issue resolved now" (see https://github.com/sarugaku/resolvelib/issues/69#issuecomment-927333533).

Replace `resolvelib` version upper cap with `resolvelib>=0.8.1,<0.9.0` with corresponding `find_matches()` and `get_preference()` interface change should now be good enough.

Signed-off-by: Wong Hoi Sing Edison <hswong3i@pantarei-design.com>

Ref: https://github.com/ansible/ansible/issues/74569

##### ISSUE TYPE

- Feature Pull Request

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -7,7 +7,4 @@ jinja2 >= 3.0.0
 PyYAML
 cryptography
 packaging
-# NOTE: resolvelib 0.x version bumps should be considered major/breaking
-# NOTE: and we should update the upper cap with care, at least until 1.0
-# NOTE: Ref: https://github.com/sarugaku/resolvelib/issues/69
-resolvelib >= 0.5.3, < 0.6.0  # dependency resolver used by ansible-galaxy
+resolvelib >= 0.8.1
```
Comment (webknjaz) [inline requirements.txt]: I think this should be
```suggestion
# NOTE: resolvelib 0.x version bumps should be considered major/breaking
# NOTE: and we should update the upper cap with care, at least until 1.0
# NOTE: Ref: https://github.com/sarugaku/resolvelib/issues/69
resolvelib >= 0.5.3, < 0.9.0  # dependency resolver used by ansible-galaxy
```

File: test/lib/ansible_test/_data/requirements/ansible.txt [hunk 2]
```diff
@@ -7,7 +7,4 @@ jinja2 >= 3.0.0
 PyYAML
 cryptography
 packaging
-# NOTE: resolvelib 0.x version bumps should be considered major/breaking
-# NOTE: and we should update the upper cap with care, at least until 1.0
-# NOTE: Ref: https://github.com/sarugaku/resolvelib/issues/69
-resolvelib >= 0.5.3, < 0.6.0  # dependency resolver used by ansible-galaxy
+resolvelib >= 0.8.1
```
Comment (webknjaz) [inline test/lib/ansible_test/_data/requirements/ansible.txt]:
```suggestion
# NOTE: resolvelib 0.x version bumps should be considered major/breaking
# NOTE: and we should update the upper cap with care, at least until 1.0
# NOTE: Ref: https://github.com/sarugaku/resolvelib/issues/69
resolvelib >= 0.5.3, < 0.9.0  # dependency resolver used by ansible-galaxy
```

**Review by webknjaz — CHANGES_REQUESTED** [review]
This change is incomplete and requires more care.
1) We shouldn't bump the minimum bound without a serious reason to do so, different downstream may be using an older version in their ecosystems / distro repos.
2) This must be tested for each of 0.5.x, 0.6.x, 0.7.x, and 0.8.x, at least manually.
3) Compared to 0.5.x, 0.6.x has changed the API (the provider class method signatures). In order to upgrade further, somebody must implement those changes while keeping backward compatibility in mind. Plus, I haven't changed how/if the later versions changed the API.
   This is also mentioned in https://github.com/ansible/ansible/issues/74569.
4) Don't remove the comments, they still apply. Just because an issue is marked as resolved it doesn't mean that it solves the current concern. I haven't seen any commitment announcements to follow some versioning scheme with predictable API breakages.

**Comment by hswong3i** [comment]
@samdoran both `find_matches()` and `get_preference()` got updated for 0.8.1, so here I just bump version dependency into `resolvelib>=0.8.1,<0.9.0`, similar as that for pip with https://github.com/pypa/pip/commit/93f5bf702c82037ec5945278be3684418326a84f

**Comment by hswong3i** [comment]
@samdoran so CI passed now ;-)

**Comment by hswong3i** [comment]
Moreover, it is now working for my Ansible deb/rpm packaging, see:
* https://build.opensuse.org/package/show/home:alvistack/ansible-ansible-2.12.0
* https://github.com/alvistack/ansible-ansible/commit/ef2e287d2d2e34390c74086e13c45daf472dd13f

File: lib/ansible/galaxy/dependency_resolution/providers.py [hunk 3]
```diff
@@ -13,13 +13,6 @@
 except ImportError:
     TYPE_CHECKING = False
 
-if TYPE_CHECKING:
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/providers.py]: Why did you move this block? It's best not to touch parts that are not really related. Maybe you could separate this into two if-blocks?

File: lib/ansible/galaxy/dependency_resolution/providers.py [hunk 4]
```diff
@@ -126,9 +129,11 @@ def identify(self, requirement_or_candidate):
 
     def get_preference(
             self,  # type: CollectionDependencyProvider
-            resolution,  # type: Optional[Candidate]
-            candidates,  # type: List[Candidate]
-            information,  # type: List[NamedTuple]
+            identifier,  # type: str
+            resolutions,  # type: Mapping[str, Candidate]
+            candidates,  # type: Mapping[str, Iterator[Candidate]]
+            information,  # type: Mapping[str, Iterable[PreferenceInformation]]
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/providers.py]: This is a breaking change that will make the downstream packaging hard. The API compatibility should be with all the minor versions stated in the previous review.

File: lib/ansible/galaxy/dependency_resolution/providers.py [hunk 5]
```diff
@@ -182,7 +188,7 @@ def get_preference(
             return float('-inf')
         return len(candidates)
 
-    def find_matches(self, requirements):
+    def find_matches(self, identifier, requirements, incompatibilities):
```
Comment (webknjaz) [inline lib/ansible/galaxy/dependency_resolution/providers.py]: The new implementation should support signatures for a number of resolvelib versions. They could be defined dynamically in runtime.

**Review by webknjaz — CHANGES_REQUESTED** [review]
Most of the points from the previous review have been ignored. I'll reiterate the main one: do not change the lower bound. You can read the rest in my previous comments.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 3 above)
Comment (hswong3i) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: Because pylint complain about that if keeping original ordering.

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 4 above)
Comment (hswong3i) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: resolvelib is a relatively new library, which not much distribution are now having official package for it:
* https://pkgs.org/download/python3-resolvelib
* Only Ubuntu 21.04+, Debian 11+, openSUSE Leap 15.2+, Fedora 34+

Moreover, if ansible 2.12.0 already drop Python support <3.7, I could see most distributions (e.g. CentOS 7, CentOS 8, Debian 10, openSUSE Leap 15.2, openSUSE Leap 15.3, Ubuntu 18.04, etc) are no longer supported.

In this case, the most workable installation method for ansible 2.12.0+ would be:
* Traditional `pip install ansible` or `pipx install ansible` which goes to `/usr/local/lib` (therefore, nothing about downstream packaging difficulties), or
* Someone crazily packaging ALL related packages with latest compatible version deb/rpm (which go for something like my https://build.opensuse.org/project/show/home:alvistack solution), or
* Modern upcoming distribution bundling Python 3.8+, resolvelib 0.8.1+, ansible 2.12.0+ all together (in the future release)

For either 3 cases we have no strong reason to pin resolvelib <0.6.0, isn't it?

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 5 above)
Comment (hswong3i) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: Refer to my above feedback: if ansible 2.12.0 is target for Python 3.8+ (which brake most legacy 2.5+ years ago distributions), keeping a multiple support for resolvelib 0.6.0/0.7.0/0.8.0/etc are most likely pointless?

Similar choose as like as pip: they also switch to using resolvelib, and supporting the latest version API interface (i.e. 0.8.1 now), with a single implementation.

Downstream packaging my pin to corresponding release according to multiple reason, e.g. different packages sharing a single legacy version of library. But this should be a reason for upstream keep pinning with a legacy version, isn't it?

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 4 above)
Comment (sivel) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: > Moreover, if ansible 2.12.0 already drop Python support <3.7, I could see most distributions (e.g. CentOS 7, CentOS 8, Debian 10, openSUSE Leap 15.2, openSUSE Leap 15.3, Ubuntu 18.04, etc) are no longer supported.

That would be an incorrect assumption.  While those distros may have a default python lower than 3.8, many of them provide newer python versions that can still be installed.

Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: You missed our main downstream which is RHEL. It's usually non-trivial to get packages updated there. Plus I'm not sure if Ubuntu PPAs have caught up. And what about the rolling-release distros like Gentoo/Arch where some software may still need an older version, for example?

Ansible is not pip, we don't vendor resolvelib. This is why we cannot just go and drop the lower bounds. Historically, we've always kept high compatibility with older systems. I've requested more reviews to gather opinions.

Comment (sivel) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: I am in agreement with @webknjaz 

Please keep the lower bounds of the original requirements.

Comment (hswong3i) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: > That would be an incorrect assumption. While those distros may have a default python lower than 3.8, many of them provide newer python versions that can still be installed.

* CentOS 7: SCL? https://centos.pkgs.org/7/centos-sclo-rh-x86_64/rh-python38-2.0-4.el7.x86_64.rpm.html
* CentOS 8: AppStream? https://centos.pkgs.org/8-stream/centos-appstream-x86_64/python38-3.8.12-1.module_el8.6.0+929+89303463.x86_64.rpm.html
* Ubuntu 18.04: PPA? https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa
* Debian 10: Personal repo? https://people.debian.org/~paravoid/python-all/

For neither above cases, you couldn't find resolvelib prepackaged (see https://pkgs.org/download/python3-resolvelib) for ansible 2.12.0+

If we could not even find a resolvelib <0.6.0 official deb/rpm with above legacy 2.5+ years ago distribution, we need someone package it as deb/rpm, or people need to install it with pip/pipx. 

P.S. this is also my root cause for packaging EVERYTHING ansible related dependency wih latest version into deb/rpm, for ALL of my support OS, see https://build.opensuse.org/project/show/home:alvistack

Comment (hswong3i) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: > You missed our main downstream which is RHEL. It's usually non-trivial to get packages updated there. Plus I'm not sure if Ubuntu PPAs have caught up. And what about the rolling-release distros like Gentoo/Arch where some software may still need an older version, for example?

When ansible 2.12.0+ choose to drop Python <3.7 support, our main downstream RHEL 7/8 could only pin with ansible 2.11.6+ (unless these downstream also packaging both Python 3.8 + all python dependencies into an official supported repo); for RHEL 9 beta it is Python 3.9 which shouldn't have difficulties for both ansible 2.12.0+ and resolvelib 0.8.1+?

File: lib/ansible/galaxy/dependency_resolution/providers.py (hunk 5 above)
Comment (webknjaz) [reply lib/ansible/galaxy/dependency_resolution/providers.py]: It is a reason to keep the API compatible.

**Comment by ssbarnea** [comment]
@hswong3i @webknjaz Any chance to revive work on this? We should really allow Ansible to use newer resolvelib, and if possible without removing support for older ones. I mention this because it would find unlikely for already release distros to want to upgrade resolvelib. They will likely only allow patch level fixes but not major new versions (Sadly 0.8 counts as major due to being under 1.0).

**Comment by hswong3i** [comment]
Changes:
* PR updated for v2.13.0
* Applied to https://github.com/alvistack/ansible-ansible/commit/68fb8d6c93a34c444bb62a7d6e9e7637e1ba1f1d
* Packaged with OBS as https://build.opensuse.org/package/show/home:alvistack/ansible-ansible-2.13.0

**Comment by ssbarnea** [comment]
Why we are not vendoring resolvelib instead, so we can avoid conflicts with various versions?

**Comment by hswong3i** [comment]
> Why we are not vendoring resolvelib instead, so we can avoid conflicts with various versions?

Umm... Good question, but I am not a suitable person for answering... I am just a packager who hope to simplify my daily deployment difficulty with on-time ad-hoc bugfix, and that why for this PR :-P

**Comment by s-hertel** [comment]
@hswong3i @ssbarnea The issues pointed out by the reviews have not been fixed - this is still a breaking change. I have a pull request https://github.com/ansible/ansible/pull/77649 to support resolvelib>=05.3,<0.9.0. It's pretty much ready to merge, so please test it out and leave feedback if you want. It's just waiting on https://github.com/ansible/ansible/pull/77630, which I would like to get in first so I can easily backport it, since it shares a bit of the code.

**Comment by hswong3i** [comment]
From supply and demand point of view:
1. We are requiring Python >= 3.8 for Ansible > 5.0
2. From https://pkgs.org/search/?q=python3-resolvelib even CentOS 9 Stream is now bundling with 0.5.4, Fedora Rawhide with 0.5.5

Please kindly understand that OS packagers are ALWAYS waiting for upstream upgrade its dependency (demand), and so they package a newer version for making it works (supply).

On the other hand we had request resolvelib for a 1.0 release around 6 months, but none of new 0.8.6+ pre-release or 1.0+ released. It is now relatively stable and so changes is not such active, too.

IMHO supporting >= 0.5.4, < 9.0 for a non 1.0+ release would be something too greedy and over complicated.

Shouldn't we keep it simple stupid: 
1. we as upstream update our package dependency to latest stable (>= 8.1.0, < 0.9), 
2. downstream OS packager adopt the change and so package resolvelib 0.8.6 if they also hope to provide a newer version of Ansible for their user, 
3. for active developer using virtualenv or pipx to manage the changes independently?
